core/src/storage/traits/push.rs

```diff
@@ -0,0 +1,304 @@
+                }
+            }
+        )*
+    }
+}
+forward_supported_array_lens!(impl_push_for_array);
+
+macro_rules! impl_push_tuple {
+    ( $($frag:ident),* $(,)? ) => {
+        impl<$($frag),*> PushForward for ($($frag),* ,)
+        where
+            $(
+                $frag: PushForward,
+            )*
+        {
+            fn push_forward(&self, ptr: &mut KeyPtr) {
+                #[allow(non_snake_case)]
+                let ($($frag),*,) = self;
+                $(
+                    PushForward::push_forward($frag, ptr);
+                )*
+            }
+        }
+
+        impl<$($frag),*> PushAt for ($($frag),* ,)
+        where
+            Self: scale::Encode,
+        {
+            fn push_at(&self, at: Key) {
+                env::set_contract_storage::<Self>(at, self)
+            }
+        }
+    }
+}
+impl_push_tuple!(A);
+impl_push_tuple!(A, B);
+impl_push_tuple!(A, B, C);
+impl_push_tuple!(A, B, C, D);
+impl_push_tuple!(A, B, C, D, E);
+impl_push_tuple!(A, B, C, D, E, F);
+impl_push_tuple!(A, B, C, D, E, F, G);
+impl_push_tuple!(A, B, C, D, E, F, G, H);
+impl_push_tuple!(A, B, C, D, E, F, G, H, I);
+impl_push_tuple!(A, B, C, D, E, F, G, H, I, J);
+
+impl PushForward for () {
+    fn push_forward(&self, _ptr: &mut KeyPtr) {}
+}
+
+impl PushAt for () {
+    fn push_at(&self, _at: Key) {}
+}
+
+impl<T> PushForward for PhantomData<T> {
+    fn push_forward(&self, _ptr: &mut KeyPtr) {}
+}
+
+impl<T> PushAt for PhantomData<T> {
+    fn push_at(&self, _at: Key) {}
+}
+
+impl<T> PushForward for Option<T>
+where
+    T: PushForward + StorageSize,
+{
+    /// We implement `PushForward` for `Option<T>` in an optimized fashion
+    /// leaving behind a cleared contract storage cell area in case of `None`.
+    ///
+    /// This can be tricky and a performance hazard for types that occupy
+    /// a large amount of storage cells (i.e. their `StorageSize::SIZE` is
+    /// big.). This implementation needs protection against this sort of hazard.
+    fn push_forward(&self, ptr: &mut KeyPtr) {
+        match self {
+            Some(val) => <T as PushForward>::push_forward(val, ptr),
+            None => {
+                // We still need to advance the key pointer.
+                let pos0 = ptr.next_for::<T>();
+                // Bail out early if `StorageSize` is too big and the method
+                // is used even though we have tried to prevent this at compile
+                // time.
+                if <T as StorageSize>::SIZE > 32 {
+                    return
+                }
+                // In theory we'd need to clear all subfields of the `Option<T>`
+                // which are technically `<T as StorageSize>::SIZE` many.
+                // However, this could be a performance hazard so we should
+                // really not do that. We should add a check to disable this
+                // implementation for `StorageSize` of more than 32. (TODO)
```

Comment:
TODO in future or as part of this PR?

Reply:
That's a good question. I hope I can find a proper solution before we merge this PR.